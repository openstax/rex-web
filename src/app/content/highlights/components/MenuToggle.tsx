import { DotMenuIcon, DotMenuToggle } from '../../../components/DotMenu';

// tslint:disable-next-line:variable-name
export const MenuIcon = DotMenuIcon;
export default DotMenuToggle;
