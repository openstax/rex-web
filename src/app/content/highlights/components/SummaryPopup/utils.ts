import { Highlight } from '@openstax/highlighter/dist/api';
import queryString from 'querystring';
import React from 'react';
import { useSelector } from 'react-redux';
import { systemQueryParameters } from '../../../../navigation/selectors';
import { Book } from '../../../types';
import { findArchiveTreeNodeById } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const useCreateHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const [link, setLink] = React.useState('');
  const systemQueryParams = useSelector(systemQueryParameters);

  React.useEffect(() => {
    const page = book ? findArchiveTreeNodeById(book.tree, sourceId) : undefined;
    const data: Record<string, string> = {
      ...systemQueryParams,
      target: JSON.stringify({ id, type: 'highlight' }),
    };
    const target = queryString.stringify(data) + `#${elementId}`;
    setLink(`${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`);
  }, [id, book, elementId, sourceId, systemQueryParams]);

  return link;
};
