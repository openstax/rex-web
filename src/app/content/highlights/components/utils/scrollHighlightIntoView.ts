import { Highlight } from '@openstax/highlighter';
import { HTMLElement } from '@openstax/types/lib.dom';
import { scrollIntoView } from '../../../../domUtils';

export default (highlight: Highlight, cardRef?: React.RefObject<HTMLElement>) => {
  const firstElement = highlight.elements[0] as HTMLElement;
  const lastElement = highlight.elements[highlight.elements.length - 1] as HTMLElement;
  const otherElements = [lastElement];
  if (cardRef?.current) {
    otherElements.push(cardRef.current);
  }
  scrollIntoView(firstElement, otherElements);
};
