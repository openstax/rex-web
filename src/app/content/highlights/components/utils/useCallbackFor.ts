import React from 'react';

// Some functions get recreated but shouldn't be reacted to
// This makes a non-changing callback
export default function useCallbackFor<T extends Array<unknown>>(fn: (...args: T) => unknown) {
    const ref = React.useRef(fn);
    const cb = React.useCallback(
        (...args: T) => ref.current(...args),
        []
    );

    ref.current = fn;
    return cb;
}
