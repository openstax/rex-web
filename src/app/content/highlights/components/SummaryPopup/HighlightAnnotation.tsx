import { HTMLTextAreaElement } from '@openstax/types/lib.dom';
import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import styled from 'styled-components/macro';
import Button from '../../../../components/Button';
import { textRegularStyle } from '../../../../components/Typography';
import theme from '../../../../theme';
import { HighlightEditButtons } from './styles';

// tslint:disable-next-line:variable-name
const HighlightNote = styled.div`
  ${textRegularStyle}
  padding-top: 1.2rem;
  display: flex;

  span {
    margin: 0 0.8rem 0 0;
    overflow: visible;
  }
`;

// tslint:disable-next-line:variable-name
const HighlightNoteAnnotation = styled.div`
  white-space: pre-wrap;
  word-break: break-word;
  overflow-wrap: break-word;
`;

// tslint:disable-next-line:variable-name
const Textarea = styled.textarea`
  ${textRegularStyle}
  font-family: inherit;
  flex: 1;
  letter-spacing: 0;
  line-height: 20px;
  color: ${theme.color.text.label};
  padding: 8px;
`;

interface HighlightAnnotationProps {
  annotation: string;
  isEditing: boolean;
  onSave: (annotation: string) => void;
  onCancel: () => void;
}

// tslint:disable-next-line:variable-name
const HighlightAnnotation = (
  { annotation, isEditing, onSave, onCancel }: HighlightAnnotationProps
) => {
  const [anno, setAnno] = React.useState(annotation);
  const intl = useIntl();

  React.useEffect(() => {
    setAnno(annotation);
  }, [annotation]);

  if (anno.length === 0 && !isEditing) { return null; }

  return <HighlightNote>
    {isEditing
      ? <Textarea
        value={anno}
        placeholder={intl.formatMessage({id: 'i18n:highlighting:card:placeholder'})}
        autoFocus={true}
        aria-label={intl.formatMessage({id: 'i18n:highlighting:card:aria-label'})}
        onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => {
          setAnno(e.target.value);
        }}
      />
      : <React.Fragment>
        <span className='highlight-note-text'>
          <FormattedMessage id='i18n:toolbar:highlights:popup:body:note:text'>
            {(msg) => msg}
          </FormattedMessage>
        </span>
        <HighlightNoteAnnotation>
          {annotation}
        </HighlightNoteAnnotation>
      </React.Fragment>
    }
    {isEditing && <HighlightEditButtons>
      <Button
        data-testid='save'
        data-analytics-label='save'
        size='medium'
        variant='primary'
        aria-label={intl.formatMessage({id: 'i18n:highlighting:button:save'})}
        onClick={() => onSave(anno)}
      >{intl.formatMessage({id: 'i18n:highlighting:button:save'})}</Button>
      <Button
        size='medium'
        data-analytics-label='cancel'
        data-testid='cancel'
        aria-label={intl.formatMessage({id: 'i18n:highlighting:button:cancel'})}
        onClick={() => {
          onCancel();
          setAnno(annotation);
        }}
      >{intl.formatMessage({id: 'i18n:highlighting:button:cancel'})}</Button>
    </HighlightEditButtons>}
  </HighlightNote>;
};

export default HighlightAnnotation;
