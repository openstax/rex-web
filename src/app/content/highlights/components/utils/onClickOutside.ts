import { Element, EventListenerOptions, HTMLElement, MouseEvent } from '@openstax/types/lib.dom';
import curry from 'lodash/fp/curry';
import React from 'react';
import { isHtmlElement } from '../../../../guards';
import { assertDocument, assertWindow } from '../../../../utils';

const targetBelongsToElement = curry((event: MouseEvent, el: React.RefObject<HTMLElement> | HTMLElement): boolean =>
  isHtmlElement(el)
    ? el.contains(event.target as Element)
    : el.current
      ? el.current.contains(event.target as Element)
      : false
);

const onClickOutside = (
  element: React.RefObject<HTMLElement> | Array<React.RefObject<HTMLElement> | HTMLElement>,
  isEnabled: boolean,
  cb: (ev: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => () => {
  if (typeof document === 'undefined') {
    return () => null;
  }

  const ifOutside = (e: MouseEvent) => {
    if (!(e.target instanceof assertWindow().Element)) {
      return;
    } else if (Array.isArray(element) && element.some(targetBelongsToElement(e))) {
      return;
    } else if (!Array.isArray(element) && targetBelongsToElement(e, element)) {
      return;
    }
    cb(e);
  };

  if (isEnabled) {
    document.addEventListener('click', ifOutside, eventOptions);
  }

  return () => assertDocument().removeEventListener('click', ifOutside, eventOptions);
};

export const useOnClickOutside = (
  element: React.RefObject<HTMLElement> | Array<React.RefObject<HTMLElement> | HTMLElement>,
  isEnabled: boolean,
  cb: (e: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  React.useEffect(onClickOutside(element, isEnabled, cb, eventOptions), [element, cb, isEnabled]);
};

export default onClickOutside;

const isRefWithHtmlElement = (el: any): el is React.RefObject<HTMLElement> => {
  return el instanceof Object && isHtmlElement(el.current);
};

export const isElementForOnClickOutside = (el: any): el is HTMLElement | React.RefObject<HTMLElement> => {
  return isHtmlElement(el) || isRefWithHtmlElement(el);
};
