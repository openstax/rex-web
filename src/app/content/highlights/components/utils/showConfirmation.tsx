import React from 'react';
import ReactDOM from 'react-dom';
import { RawIntlProvider } from 'react-intl';
import uuid from 'uuid/v4';
import { AppServices } from '../../../../types';
import { assertDocument, assertNotNull } from '../../../../utils';
import ConfirmationModal from '../ConfirmationModal';

export default async(services: AppServices) => {
  const document = assertDocument();
  const domNode = document.createElement('div');

  domNode.id = `dialog-${uuid()}`;
  const root = assertNotNull(document.getElementById('root'), 'root element not found');
  root.insertAdjacentElement('afterend', domNode);

  const userChoice: boolean = await new Promise((resolve) => {
    const confirm = () => resolve(true);
    const deny = () => resolve(false);
    ReactDOM.render(
      <RawIntlProvider value={services.intl}>
        <ConfirmationModal deny={deny} confirm={confirm} />
      </RawIntlProvider>,
      domNode
    );
  });

  ReactDOM.unmountComponentAtNode(domNode);
  assertNotNull(domNode.parentElement, 'parent element not found').removeChild(domNode);

  return userChoice;
};
