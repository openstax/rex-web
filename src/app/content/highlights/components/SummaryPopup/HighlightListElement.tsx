import { HighlightColorEnum, HighlightUpdateColorEnum } from '@openstax/highlighter/dist/api';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import styled, { css } from 'styled-components/macro';
import { useAnalyticsEvent } from '../../../../../helpers/analytics';
import theme from '../../../../theme';
import ContentExcerpt from '../../../components/ContentExcerpt';
import { highlightStyles } from '../../../constants';
import { book as bookSelector } from '../../../selectors';
import { popupBodyPadding } from '../../../styles/PopupStyles';
import { requestDeleteHighlight, updateHighlight } from '../../actions';
import { HighlightData } from '../../types';
import ContextMenu from './ContextMenu';
import HighlightAnnotation from './HighlightAnnotation';
import HighlightDeleteWrapper from './HighlightDeleteWrapper';
import { useCreateHighlightLink } from './utils';

// tslint:disable-next-line:variable-name
const HighlightOuterWrapper = styled.div`
  position: relative;
  overflow: visible;

  :not(:last-child) {
    border-bottom: solid 0.2rem ${theme.color.neutral.darker};
  }

  background: ${theme.color.neutral.base};

  @media print {
    border-width: 0;
    position: relative;
    page-break-inside: avoid;
    background: white;
  }
`;

// tslint:disable-next-line:variable-name
export const HighlightContentWrapper = styled.div`
  padding: 1.2rem ${popupBodyPadding}rem;
  ${(props: {color: string}) => {
    const style = highlightStyles.find((search) => search.label === props.color);

    if (!style) {
      return null;
    }

    return css`
      border-left: solid 0.8rem ${style.focused};

      ${ContentExcerpt} {
        background-color: ${style.passive};
      }

      .highlight-note-text {
        color: ${style.focused};
      }
    `;
  }}

  @media print {
    break-inside: avoid-page;

    ${ContentExcerpt} {
      background-color: white;
    }
  }
`;

interface HighlightListElementProps {
  highlight: HighlightData;
  locationFilterId: string;
  pageId: string;
}

// tslint:disable-next-line:variable-name
const HighlightListElement = ({ highlight, locationFilterId, pageId }: HighlightListElementProps) => {
  const [isEditing, setIsEditing] = React.useState(false);
  const [isDeleting, setIsDeleting] = React.useState(false);
  const book = useSelector(bookSelector);
  const dispatch = useDispatch();
  const linkToHighlight = useCreateHighlightLink(highlight, book);

  const trackEditNoteColor = useAnalyticsEvent('editNoteColor');
  const trackEditAnnotation = useAnalyticsEvent('editAnnotation');
  const trackDeleteHighlight = useAnalyticsEvent('deleteHighlight');

  const updateAnnotation = (
    annotation: string
  ) => {
    const addedNote = (highlight.annotation === undefined);

    dispatch(updateHighlight({
      highlight: {annotation},
      id: highlight.id,
    }, {
      locationFilterId,
      pageId,
      preUpdateData: {
        highlight: {
          annotation: highlight.annotation,
          color: highlight.color as string as HighlightUpdateColorEnum,
        },
        id: highlight.id,
      },
    }));
    trackEditAnnotation(addedNote, highlight.color, true);
    setIsEditing(false);
  };

  const updateColor = (color: HighlightColorEnum) => {
    dispatch(updateHighlight({
      highlight: {color: color as string as HighlightUpdateColorEnum},
      id: highlight.id,
    }, {
      locationFilterId,
      pageId,
      preUpdateData: {
        highlight: {
          annotation: highlight.annotation,
          color: highlight.color as string as HighlightUpdateColorEnum,
        },
        id: highlight.id,
      },
    }));
    trackEditNoteColor(color, true);
  };

  const confirmDelete = () => {
    dispatch(requestDeleteHighlight(highlight, {
      locationFilterId,
      pageId,
    }));
    trackDeleteHighlight(highlight.color, true);
  };

  return <HighlightOuterWrapper>
    {!isEditing && <ContextMenu
      highlight={highlight}
      linkToHighlight={linkToHighlight}
      onDelete={() => setIsDeleting(true)}
      onEdit={() => setIsEditing(true)}
      onColorChange={updateColor}
    />}
    <HighlightContentWrapper color={highlight.color}>
      <ContentExcerpt
        data-highlight-id={highlight.id}
        content={highlight.highlightedContent}
        source={highlight.sourceId}
      />
      <HighlightAnnotation
        annotation={highlight.annotation || ''}
        isEditing={isEditing}
        onSave={updateAnnotation}
        onCancel={() => setIsEditing(false)}
      />
    </HighlightContentWrapper>
    {isDeleting && <HighlightDeleteWrapper
      hasAnnotation={Boolean(highlight.annotation)}
      onCancel={() => setIsDeleting(false)}
      onDelete={confirmDelete}
    />}
  </HighlightOuterWrapper>;
};

export default HighlightListElement;
