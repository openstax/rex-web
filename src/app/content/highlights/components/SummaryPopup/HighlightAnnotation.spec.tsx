import React from 'react';
import renderer from 'react-test-renderer';
import createTestServices from '../../../../../test/createTestServices';
import createTestStore from '../../../../../test/createTestStore';
import { book, page } from '../../../../../test/mocks/archiveLoader';
import createMockHighlight from '../../../../../test/mocks/highlight';
import { mockCmsBook } from '../../../../../test/mocks/osWebLoader';
import TestContainer from '../../../../../test/TestContainer';
import { Store } from '../../../../types';
import { receiveBook, receivePage } from '../../../actions';
import { formatBookData } from '../../../utils';
import { highlightLocationFilters } from '../../selectors';
import { HighlightData } from '../../types';
import { getHighlightLocationFilterForPage } from '../../utils';
import MenuToggle from '../MenuToggle';
import HighlightAnnotation from './HighlightAnnotation';
import HighlightListElement from './HighlightListElement';
import * as utils from './utils';

describe('HighlightDeleteWrapper', () => {
  let store: Store;

  beforeEach(() => {
    store = createTestStore();
  });

  it('match snapshot when editing is set to false', () => {
    const component = renderer.create(<TestContainer store={store}>
      <HighlightAnnotation
        annotation='Some annotation'
        isEditing={false}
        // tslint:disable-next-line: no-empty
        onSave={() => {}}
        // tslint:disable-next-line: no-empty
        onCancel={() => {}}
      />
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('match snapshot when editing is set to true', () => {
    const component = renderer.create(<TestContainer store={store}>
      <HighlightAnnotation
        annotation='Some annotation'
        isEditing={true}
        // tslint:disable-next-line: no-empty
        onSave={() => {}}
        // tslint:disable-next-line: no-empty
        onCancel={() => {}}
      />
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('properly fire onCancel and onSave props', () => {
    let savedText = '';
    let cancelClicked = false;
    const annotation = 'TEST';

    const component = renderer.create(<TestContainer store={store}>
      <HighlightAnnotation
        annotation={annotation}
        isEditing={true}
        onCancel={() => { cancelClicked = true; }}
        onSave={(text) => { savedText = text; }}
      />
    </TestContainer>);

    renderer.act(() => { return; });

    renderer.act(() => {
      const textarea = component.root.findByProps({ value: annotation });
      textarea.props.onChange({
        preventDefault: jest.fn(),
        target: {value: 'newAnno'},
      });
    });

    renderer.act(() => {
      const saveButton = component.root.findByProps({ 'data-testid': 'save' });
      saveButton.props.onClick();
      const cancelButton = component.root.findByProps({ 'data-testid': 'cancel' });
      cancelButton.props.onClick();
    });

    expect(savedText).toEqual('newAnno');
    expect(cancelClicked).toEqual(true);
  });
});

describe('Highlight annotation', () => {
  let store: Store;
  let services: ReturnType<typeof createTestServices>;
  let highlight: ReturnType<typeof createMockHighlight>;

  beforeEach(() => {
    store = createTestStore();
    services = createTestServices();
    highlight = createMockHighlight('asdf');
  });

  it('edits annotation from MH pop up and track analytics', () => {
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));
    store.dispatch(receivePage({...page, references: []}));

    const locationFilters = highlightLocationFilters(store.getState());
    const location = getHighlightLocationFilterForPage(locationFilters, page);

    jest.spyOn(utils, 'useCreateHighlightLink')
      .mockReturnValue('/link/to/highlight');

    const component = renderer.create(<TestContainer services={services} store={store}>
      <HighlightListElement
        highlight={highlight as unknown as HighlightData}
        locationFilterId={location!.id}
        pageId={page.id}
      />
    </TestContainer>);

    const track = jest.spyOn(services.analytics.editAnnotation, 'track');

    renderer.act(() => {
      const dropdown = component.root.findByType(MenuToggle);
      dropdown.props.onClick();
    });

    renderer.act(() => {
      const editButton = component.root.findByProps({ 'data-testid': 'edit' });
      editButton.props.onClick();
    });

    renderer.act(() => {
      const textarea = component.root.findByType('textarea');
      textarea.props.onChange({
        preventDefault: jest.fn(),
        target: {value: 'newAnno'},
      });
    });

    renderer.act(() => {
      const saveButton = component.root.findByProps({ 'data-testid': 'save' });
      saveButton.props.onClick();
    });

    expect(track).toHaveBeenCalled();
  });
});
