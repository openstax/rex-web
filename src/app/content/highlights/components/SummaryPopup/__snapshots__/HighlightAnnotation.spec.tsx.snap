// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightDeleteWrapper match snapshot when editing is set to false 1`] = `
<div
  className="HighlightAnnotation__HighlightNote-ppiq8t-0 fKLraL"
>
  <span
    className="highlight-note-text"
  >
    Note:
  </span>
  <div
    className="HighlightAnnotation__HighlightNoteAnnotation-ppiq8t-1 cWfCT"
  >
    Some annotation
  </div>
</div>
`;

exports[`HighlightDeleteWrapper match snapshot when editing is set to true 1`] = `
<div
  className="HighlightAnnotation__HighlightNote-ppiq8t-0 fKLraL"
>
  <textarea
    aria-label="Enter note's text"
    autoFocus={true}
    className="HighlightAnnotation__Textarea-ppiq8t-2 kebluY"
    onChange={[Function]}
    placeholder="Add a note..."
    value="Some annotation"
  />
  <div
    className="styles__HighlightEditButtons-sc-1t1yajj-0 JvKbW"
  >
    <button
      aria-label="Save"
      className="Button-ayg7nk-0 fFJxez"
      data-analytics-label="save"
      data-testid="save"
      onClick={[Function]}
    >
      Save
    </button>
    <button
      aria-label="Cancel"
      className="Button-ayg7nk-0 liRzrl"
      data-analytics-label="cancel"
      data-testid="cancel"
      onClick={[Function]}
    >
      Cancel
    </button>
  </div>
</div>
`;
